### PR TITLE
fix(grouping): #MVS-72 fix grouping controller and services

### PR DESCRIPTION
**Pull request by BastienBenard**
## Describe your changes
Fix back to answer api doc

Unit tests are currently not passing due to a simulation of successful requests in the service.

Doc: https://confluence.support-ent.fr/display/MTVS/Contrat+d%27API#Contratd'API-addGroupingAudience

## Issue ticket number and link
[MVS-72](https://jira.support-ent.fr/browse/MVS-72)

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

**Review by vava56754 — APPROVED** [review]

File: src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupeService.java [hunk 1]
```diff
@@ -211,6 +212,21 @@ public void getNameOfGroupeClasse(String idGroupe, Handler<Either<String, JsonAr
         neo4j.execute(query, values, Neo4jResult.validResultHandler(handler));
     }
 
+    @Override
+    public Future<JsonArray> getNameOfGroupeClasse(String[] idsAudience) {
+        Promise<JsonArray> promise = Promise.promise();
+        JsonObject values = new JsonObject().put("idsAudience", new fr.wseduc.webutils.collections.JsonArray(Arrays.asList(idsAudience)));
```
Comment (flmartineau) [inline src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupeService.java:218]: pas possible de faire new JsonArray ?

File: src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupingService.java [hunk 2]
```diff
@@ -27,11 +28,87 @@ public DefaultGroupingService(ServiceFactory serviceFactory) {
...
+
+    @SuppressWarnings("unchecked")
+    private Future<JsonArray> generateGroupingsList(JsonArray groupingSQLResults) {
+        Promise<JsonArray> promise = Promise.promise();
+
+        List<String> listAudienceId = groupingSQLResults.stream().map(JsonObject.class::cast)
+                .map(jsonObject -> jsonObject.getJsonArray(Field.STUDENT_DIVISIONS).getList())
+                .map(list -> (List<JsonArray>) list)
+                .map(jsonArrays -> {
+                    return jsonArrays.stream()
```
Comment (JinDouang) [inline src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupingService.java:60]: Pas besoin du return si ce n'est qu'une ligne instruction puis tu peux avoir un lambda expression!

**Review by JinDouang — CHANGES_REQUESTED** [review]

File: src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupeService.java [hunk 3]
```diff
@@ -128,8 +129,8 @@ public void getClasseGroupe(String[] idGroupe, Handler<Either<String, JsonArray>
                 .append(" WHERE n.id IN {idGroupe} WITH  s, n, u ")
                 .append(" MATCH (c:Class)-[:BELONGS]->(s) WHERE c.externalId IN u.classes RETURN n.id as id_groupe, ")
                 .append(" COLLECT(DISTINCT c.id) AS id_classes ");
-        params.put("idGroupe", new fr.wseduc.webutils.collections.JsonArray(Arrays.asList(idGroupe)))
-                .put("idGroupe", new fr.wseduc.webutils.collections.JsonArray(Arrays.asList(idGroupe)));
+        params.put("idGroupe", new JsonArray(Arrays.asList(idGroupe)))
```
Comment (JinDouang) [inline src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupeService.java:132]: Si possible gerer le magic string

File: src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupingService.java [hunk 4]
```diff
@@ -27,11 +28,88 @@ public DefaultGroupingService(ServiceFactory serviceFactory) {
...
+    }
+
+    @SuppressWarnings("unchecked")
+    private Future<JsonArray> generateGroupingsList(JsonArray groupingSQLResults) {
+        Promise<JsonArray> promise = Promise.promise();
+
+        List<String> listAudienceId = groupingSQLResults.stream()
+                .map(JsonObject.class::cast)
+                .map(jsonObjectGrouping -> jsonObjectGrouping.getJsonArray(Field.STUDENT_DIVISIONS))
+                .map(jsonArrayStudentDivisions -> {
```
Comment (JinDouang) [inline src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupingService.java:59]: On en reparle mais tu as moyen de simplifier et de skip quelques instructions

File: src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupingService.java [hunk 5]
```diff
@@ -27,11 +28,88 @@ public DefaultGroupingService(ServiceFactory serviceFactory) {
...
+                    return jsonArrayStudentDivisions.stream()
+                            .map(JsonArray.class::cast)
+                            .map(jsonArray -> jsonArray.getString(1))
```
Comment (JinDouang) [inline src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupingService.java:62]: C'est quoi 1 ?

**Review by JinDouang — CHANGES_REQUESTED** [review]

File: src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupingService.java (hunk 5 above)
Comment (BastienBenard) [reply src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupingService.java:62]:
```
    [
       {
          "id":"505bcf09-4cb9-4294-8632-29ee5e070f5d",
          "name":"test2",
          "structure_id":"46094e4c-a86f-4b73-812e-890e791a6900",
          "student_divisions":[
             [
                1,
                "df16ac21-7677-4882-8d3d-d142950ba9db" //*1
             ]
          ]
       },
       {
          "id":"48a11c9b-b3b9-42ff-b7be-5572ed190090",
          "name":"test",
          "structure_id":"46094e4c-a86f-4b73-812e-890e791a6900",
          "student_divisions":[
             [
                1,
                "62173c9d-3bb3-44ec-9240-f30ecfb1b4b3"
             ],
             [
                2,
                "6dcc8316-1861-40fa-af52-acc8796ceb7d"
             ]
          ]
       }
    ]
```
Voici un exemple de groupingSQLResults

File: src/main/java/fr/openent/viescolaire/service/GroupeService.java [hunk 6]
```diff
@@ -62,6 +62,12 @@ void listUsersByGroupeEnseignementId(String groupeEnseignementId,String profile,
      */
     void getNameOfGroupeClasse(String idGroupe, Handler<Either<String, JsonArray>> handler);
 
+    /**
+     * @param idsAudience list of id class/group
+     * @return Future of results
+     */
+    Future<JsonArray> getNameOfGroupeClasse(List<String> idsAudience);
```
Comment (JinDouang) [inline src/main/java/fr/openent/viescolaire/service/GroupeService.java:69]: getNameOfGroupClass

File: src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupingService.java [hunk 7]
```diff
@@ -27,11 +30,85 @@ public DefaultGroupingService(ServiceFactory serviceFactory) {
...
+    @SuppressWarnings("unchecked")
+    private Future<JsonArray> generateGroupingsList(JsonArray groupingSQLResults) {
+        Promise<JsonArray> promise = Promise.promise();
+
+        List<String> listAudienceId = groupingSQLResults.stream()
+                .map(JsonObject.class::cast)
+                .map(jsonObjectGrouping -> jsonObjectGrouping.getJsonArray(Field.STUDENT_DIVISIONS))
+                .map(jsonArrayStudentDivisions -> jsonArrayStudentDivisions.stream()
+                        .map(JsonArray.class::cast)
+                        .map(jsonArrayStudentDivision -> jsonArrayStudentDivision.getString(1))
```
Comment (JinDouang) [inline src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupingService.java:63]: Pourquoi l'index 1 ?

File: src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupingService.java [hunk 8]
```diff
@@ -27,11 +30,85 @@ public DefaultGroupingService(ServiceFactory serviceFactory) {
...
+        groupService.getNameOfGroupeClasse(listAudienceId)
+                .onSuccess(jsonArray -> {
+                    Map<String, String> map = jsonArray.stream().map(JsonObject.class::cast)
+                            .collect(Collectors.toMap(jsonObject -> jsonObject.getString(Field.ID), jsonObject -> jsonObject.getString(Field.NAME)));
+                    for (Object groupingObject : groupingSQLResults) {
+                        JsonObject grouping = (JsonObject) groupingObject;
+                        JsonArray studentDivisions = grouping.getJsonArray(Field.STUDENT_DIVISIONS);
+                        List<JsonObject> jsonObjectList = (List<JsonObject>) studentDivisions.getList().stream()
+                                .map(studentDivision -> new JsonObject()
+                                        .put(Field.ID, ((JsonArray) studentDivision).getString(1))
```
Comment (JinDouang) [inline src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupingService.java:78]: Pourquoi l'index 1 ?

File: src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupingService.java [hunk 9]
```diff
@@ -27,11 +30,85 @@ public DefaultGroupingService(ServiceFactory serviceFactory) {
...
+                                        .put(Field.NAME, map.get(((JsonArray) studentDivision).getString(1)))).collect(Collectors.toList());
```
Comment (JinDouang) [inline src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupingService.java:79]: Pourquoi l'index 1 ?

File: src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupingService.java [hunk 10]
```diff
@@ -27,11 +30,85 @@ public DefaultGroupingService(ServiceFactory serviceFactory) {
...
+                        .map(JsonArray.class::cast)
+                        .map(jsonArrayStudentDivision -> jsonArrayStudentDivision.getString(1))
+                        .collect(Collectors.toList()))
+                .flatMap(List::stream)
+                .distinct()
+                .collect(Collectors.toList());
+
+        groupService.getNameOfGroupeClasse(listAudienceId)
+                .onSuccess(jsonArray -> {
+                    Map<String, String> map = jsonArray.stream().map(JsonObject.class::cast)
```
Comment (JinDouang) [inline src/main/java/fr/openent/viescolaire/service/impl/DefaultGroupingService.java:71]: Il faut revoir le nommage "map" et "jsonArray" en réponse 🤔

File: src/main/resources/sql/082-groupings.sql [hunk 11]
```diff
@@ -1,14 +1,14 @@
 CREATE TABLE viesco.grouping
 (
     id character varying(36) NOT NULL,
-    name character varying(36),
+    name character varying(255),
     structure_id character varying(36),
```
Comment (JinDouang) [inline src/main/resources/sql/082-groupings.sql:5]: structure_id, mets 255 plutôt

cf (https://github.com/OPEN-ENT-NG/competences/issues/106)

**Pull request merged**